zkevm-circuits/src/evm_circuit/execution/end_tx.rs

```diff
@@ -0,0 +1,277 @@
+            effective_refund + step.gas_left,
+            gas_fee_refund,
+        )?;
+        self.tx_caller_address
+            .assign(region, offset, tx.caller_address.to_scalar())?;
+        self.gas_fee_refund.assign(
+            region,
+            offset,
+            vec![caller_balance_pair.1, gas_fee_refund],
+            caller_balance_pair.0,
+        )?;
+        let effective_tip = tx.gas_price - block.context.base_fee;
+        self.sub_gas_price_by_base_fee.assign(
+            region,
+            offset,
+            [effective_tip, block.context.base_fee],
+            tx.gas_price,
+        )?;
+        self.mul_effective_tip_by_gas_used.assign(
+            region,
+            offset,
+            effective_tip,
+            gas_used,
+            effective_tip * gas_used,
+        )?;
+        self.coinbase
+            .assign(region, offset, block.context.coinbase.to_scalar())?;
+        self.coinbase_reward.assign(
+            region,
+            offset,
+            vec![coinbase_balance_pair.1, effective_tip * gas_used],
+            coinbase_balance_pair.0,
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{
+        test::run_test_circuit_incomplete_fixed_table, witness::block_convert,
+    };
+    use eth_types::{self, address, geth_types::Account, Address, Word};
+
+    fn test_ok(txs: Vec<eth_types::Transaction>) {
+        let accounts = txs
+            .iter()
+            .map(|tx| Account {
+                address: tx.from,
+                balance: Word::from(10).pow(20.into()),
+                ..Default::default()
+            })
+            .collect::<Vec<_>>();
+        let block_data =
+            bus_mapping::mock::BlockData::new_from_geth_data(mock::new(accounts, txs).unwrap());
+        let mut builder = block_data.new_circuit_input_builder();
+        builder
+            .handle_block(&block_data.eth_block, &block_data.geth_traces)
+            .unwrap();
+        let block = block_convert(&builder.block, &builder.code_db);
+
+        assert_eq!(run_test_circuit_incomplete_fixed_table(block), Ok(()));
+    }
+
+    fn mock_tx(from: Address, gas: Option<u64>, gas_price: Option<Word>) -> eth_types::Transaction {
+        let to = address!("0x00000000000000000000000000000000000000ff");
+        let minimal_gas = Word::from(21000);
+        let two_gwei = Word::from(2_000_000_000);
+        eth_types::Transaction {
+            from,
+            to: Some(to),
+            gas: gas.map(Word::from).unwrap_or(minimal_gas),
+            gas_price: gas_price.or(Some(two_gwei)),
+            ..Default::default()
+        }
+    }
+
+    #[test]
+    fn end_tx_gadget_simple() {
+        // Tx with non-capped refund
+        test_ok(vec![mock_tx(
+            address!("0x00000000000000000000000000000000000000fe"),
+            Some(27000),
+            None,
+        )]);
+        // TODO: Enable this when SSTORE is implemented.
```

Comment:
I don't understand why this test case is related to SSTORE.  Would this test case apply an SSTORE that causes a refund?  Where would the cap refund value be defined?

Reply:
Sorry for confusing, I should also comment the first case. My thought is, since we don't hand craft the witness anymore, `SSTORE` is the only way to make `refund` to be non-zero. So first and second testcases should have different code to execute, and second will have capped refund, first don't.

Let me fix this.